File: examples/web/src/sync/envelope.ts
/** AES-256-GCM envelope. Passphrase stays on device. */

const ITERATIONS = 210_000;

export type FeedEnvelope = {
  version: 1;
  kdf: "pbkdf2";
  aead: "aes-256-gcm";
  saltB64: string;
  nonceB64: string;
  ciphertextB64: string;
};

function asBuffer(bytes: Uint8Array): ArrayBuffer {
  return bytes.buffer.slice(bytes.byteOffset, bytes.byteOffset + bytes.byteLength) as ArrayBuffer;
}

function toB64(bytes: ArrayBuffer | Uint8Array): string {
  const view = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes);
  let bin = "";
  for (const byte of view) bin += String.fromCharCode(byte);
  return btoa(bin);
}

function fromB64(value: string): Uint8Array {
  const bin = atob(value);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i += 1) out[i] = bin.charCodeAt(i);
  return out;
}

async function deriveKey(passphrase: string, salt: Uint8Array): Promise<CryptoKey> {
  const material = await crypto.subtle.importKey(
    "raw",
    new TextEncoder().encode(passphrase),
    "PBKDF2",
    false,
    ["deriveKey"],
  );
  return crypto.subtle.deriveKey(
    { name: "PBKDF2", salt: asBuffer(salt), iterations: ITERATIONS, hash: "SHA-256" },
    material,
    { name: "AES-GCM", length: 256 },
    false,
    ["encrypt", "decrypt"],
  );
}

export async function sealJson(plain: string, passphrase: string): Promise<Uint8Array> {
  if (!passphrase) throw new Error("sync passphrase required");
  const salt = crypto.getRandomValues(new Uint8Array(16));
  const nonce = crypto.getRandomValues(new Uint8Array(12));
  const key = await deriveKey(passphrase, salt);
  const ciphertext = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv: asBuffer(nonce) },
    key,
    new TextEncoder().encode(plain),
  );
  const envelope: FeedEnvelope = {
    version: 1,
    kdf: "pbkdf2",
    aead: "aes-256-gcm",
    saltB64: toB64(salt),
    nonceB64: toB64(nonce),
    ciphertextB64: toB64(ciphertext),
  };
  return new TextEncoder().encode(JSON.stringify(envelope));
}

export async function openJson(blob: Uint8Array, passphrase: string): Promise<string> {
  if (!passphrase) throw new Error("sync passphrase required");
  const envelope = JSON.parse(new TextDecoder().decode(blob)) as FeedEnvelope;
  if (envelope.version !== 1 || envelope.aead !== "aes-256-gcm") {
    throw new Error("unsupported envelope");
  }
  const key = await deriveKey(passphrase, fromB64(envelope.saltB64));
  const plain = await crypto.subtle.decrypt(
    { name: "AES-GCM", iv: asBuffer(fromB64(envelope.nonceB64)) },
    key,
    asBuffer(fromB64(envelope.ciphertextB64)),
  );
  return new TextDecoder().decode(plain);
}

File: examples/android/app/src/main/java/org/aetherfeed/app/sync/Envelope.kt
package org.aetherfeed.app.sync

import android.util.Base64
import javax.crypto.Cipher
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec
import org.json.JSONObject

private const val ITERATIONS = 210_000

fun sealJson(plain: String, passphrase: String): ByteArray {
    require(passphrase.isNotBlank()) { "sync passphrase required" }
    val salt = randomBytes(16)
    val nonce = randomBytes(12)
    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
    cipher.init(Cipher.ENCRYPT_MODE, deriveKey(passphrase, salt), GCMParameterSpec(128, nonce))
    val ciphertext = cipher.doFinal(plain.toByteArray(Charsets.UTF_8))
    return JSONObject()
        .put("version", 1)
        .put("kdf", "pbkdf2")
        .put("aead", "aes-256-gcm")
        .put("saltB64", b64(salt))
        .put("nonceB64", b64(nonce))
        .put("ciphertextB64", b64(ciphertext))
        .toString()
        .toByteArray(Charsets.UTF_8)
}

fun openJson(blob: ByteArray, passphrase: String): String {
    require(passphrase.isNotBlank()) { "sync passphrase required" }
    val envelope = JSONObject(blob.toString(Charsets.UTF_8))
    require(envelope.optInt("version") == 1 && envelope.optString("aead") == "aes-256-gcm") {
        "unsupported envelope"
    }
    val salt = fromB64(envelope.getString("saltB64"))
    val nonce = fromB64(envelope.getString("nonceB64"))
    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
    cipher.init(Cipher.DECRYPT_MODE, deriveKey(passphrase, salt), GCMParameterSpec(128, nonce))
    return cipher.doFinal(fromB64(envelope.getString("ciphertextB64"))).toString(Charsets.UTF_8)
}

private fun deriveKey(passphrase: String, salt: ByteArray): SecretKeySpec {
    val spec = PBEKeySpec(passphrase.toCharArray(), salt, ITERATIONS, 256)
    val raw = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256").generateSecret(spec).encoded
    return SecretKeySpec(raw, "AES")
}

private fun randomBytes(size: Int): ByteArray = ByteArray(size).also { java.security.SecureRandom().nextBytes(it) }

private fun b64(bytes: ByteArray): String = Base64.encodeToString(bytes, Base64.NO_WRAP)

private fun fromB64(value: String): ByteArray = Base64.decode(value, Base64.NO_WRAP)
